File: models/donation.py
from . import db


class Donation(db.Model):
    __tablename__ = 'donations'

    id = db.Column(db.Integer, primary_key=True)
    doador_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    necessidade_id = db.Column(db.Integer, db.ForeignKey('needs.id'), nullable=False)
    tipo = db.Column(db.String(20), default='fisico')
    item = db.Column(db.String(200), nullable=False)
    quantidade = db.Column(db.String(50), nullable=False)
    categoria = db.Column(db.String(50), default='outro')
    observacao = db.Column(db.Text)
    status = db.Column(db.String(20), default='pendente')
    data_intencao = db.Column(db.DateTime, default=db.func.now())
    data_recebimento = db.Column(db.DateTime)

    doador = db.relationship('User', backref='doacoes', foreign_keys=[doador_id])
    updates = db.relationship('DonationUpdate', backref='donation', lazy='dynamic',
                              foreign_keys='DonationUpdate.donation_id',
                              order_by='DonationUpdate.data_criacao.desc()')

File: src/main/java/com/conexoessolidarias/model/Donation.java
package com.conexoessolidarias.model;

import jakarta.persistence.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "donations")
public class Donation {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "doador_id", nullable = false)
    private User doador;

    @ManyToOne
    @JoinColumn(name = "necessidade_id", nullable = false)
    private Need need;

    @Column(nullable = false)
    private String tipo = "fisico";

    @Column(nullable = false)
    private String item;

    @Column(nullable = false)
    private String quantidade;

    private String categoria = "outro";

    @Column(columnDefinition = "TEXT")
    private String observacao;

    @Column(nullable = false)
    private String status = "pendente";

    private LocalDateTime dataIntencao = LocalDateTime.now();

    private LocalDateTime dataRecebimento;

    @OneToMany(mappedBy = "donation")
    @OrderBy("dataCriacao DESC")
    private List<DonationUpdate> updates = new ArrayList<>();

    public Donation() {}

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public User getDoador() { return doador; }
    public void setDoador(User doador) { this.doador = doador; }
    public Need getNeed() { return need; }
    public void setNeed(Need need) { this.need = need; }
    public String getTipo() { return tipo; }
    public void setTipo(String tipo) { this.tipo = tipo; }
    public String getItem() { return item; }
    public void setItem(String item) { this.item = item; }
    public String getQuantidade() { return quantidade; }
    public void setQuantidade(String quantidade) { this.quantidade = quantidade; }
    public String getCategoria() { return categoria; }
    public void setCategoria(String categoria) { this.categoria = categoria; }
    public String getObservacao() { return observacao; }
    public void setObservacao(String observacao) { this.observacao = observacao; }
    public String getStatus() { return status; }
    public void setStatus(String status) { this.status = status; }
    public LocalDateTime getDataIntencao() { return dataIntencao; }
    public void setDataIntencao(LocalDateTime dataIntencao) { this.dataIntencao = dataIntencao; }
    public LocalDateTime getDataRecebimento() { return dataRecebimento; }
    public void setDataRecebimento(LocalDateTime dataRecebimento) { this.dataRecebimento = dataRecebimento; }
    public List<DonationUpdate> getUpdates() { return updates; }
    public void setUpdates(List<DonationUpdate> updates) { this.updates = updates; }
}
